### PR TITLE
Add a convention that builds a read-optimized model to be used at runtime

### LINKED ISSUE #8258: Create a read-optimized implementation of IModel
This `IModel` will be built after the model is validated and will not have the `ConfigurationSource` or other data used only during model building.

This model will be built by `IModelRuntimeInitializer` and stored in a runtime annotation in case the original model is reused.

To provide the optimized implementations default interface implementations are to be used, see https://github.com/aspnet/EntityFramework/issues/19213

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Allow the design-time model to be accessed when necessary

Fixes #8258

### CODE REVIEW

**Comment by smitpatel** [comment]
Is there any specific reason, you individually added multiple people for review rather than the team? 🤔

**Comment by AndriySvyryd** [comment]
>Is there any specific reason, you individually added multiple people for review rather than the team? 🤔
Yes, I want multiple review instead of a single one representing the team.

File: src/EFCore/Infrastructure/IModelCacheKeyFactory.cs [hunk 1]
```diff
@@ -29,6 +30,20 @@ public interface IModelCacheKeyFactory
...
+        /// </summary>
+        /// <param name="context">
+        ///     The context to get the model cache key for.
+        /// </param>
+        /// <param name="designTime"> Whether the model should contain design-time configuration.</param>
+        /// <returns> The created key. </returns>
+        object Create(DbContext context, bool designTime)
+#pragma warning disable CS0618 // Type or member is obsolete
+            => Create(context);
+#pragma warning restore CS0618 // Type or member is obsolete
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/IModelCacheKeyFactory.cs:47]: @ajcvickers

File: src/EFCore/Infrastructure/IModelRuntimeInitializer.cs [hunk 2]
```diff
@@ -29,10 +29,14 @@ public interface IModelRuntimeInitializer
...
+        /// <param name="designTime"> Whether the model should contain design-time configuration.</param>
+        /// <param name="validationLogger">
+        ///     The validation logger. If <see langword="null"/> is provided validation will not be performed.
+        ///     </param>
         /// <returns> The initialized model. </returns>
         IModel Initialize(
             IModel model,
-            IDiagnosticsLogger<DbLoggerCategory.Model.Validation>? validationLogger);
+            bool designTime = true,
+            IDiagnosticsLogger<DbLoggerCategory.Model.Validation>? validationLogger = null);
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/IModelRuntimeInitializer.cs:40]: @bricelam

File: src/EFCore/Metadata/ObjectArrayParameterBinding.cs [hunk 3]
```diff
@@ -51,5 +51,24 @@ public override Expression BindToParameter(ParameterBindingInfo bindingInfo)
...
+            var newBindings = new List<ParameterBinding>(_bindings.Count);
+            var propertyCount = 0;
+            foreach (var binding in _bindings)
+            {
+                var newBinding = binding.With(consumedProperties.Skip(propertyCount).Take(binding.ConsumedProperties.Count).ToList());
+                newBindings.Add(newBinding);
+                propertyCount += binding.ConsumedProperties.Count;
+            }
+
+            return new ObjectArrayParameterBinding(newBindings);
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/ObjectArrayParameterBinding.cs:71]: @ajcvickers

File: src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -120,7 +120,7 @@ public static class DesignTimeServiceCollectionExtensions
                 .AddTransient(_ => context.GetService<IMigrationsModelDiffer>())
                 .AddTransient(_ => context.GetService<IMigrator>())
                 .AddTransient(_ => context.GetService<IRelationalTypeMappingSource>())
-                .AddTransient(_ => context.GetService<IModel>())
+                .AddTransient(_ => context.DesignTimeModel)
```
Comment (AndriySvyryd) [inline src/EFCore.Design/Design/DesignTimeServiceCollectionExtensions.cs:123]: @bricelam @ajcvickers

File: src/EFCore.Relational/Storage/RelationalDatabaseCreator.cs [hunk 5]
```diff
@@ -144,7 +144,9 @@ public virtual Task CreateTablesAsync(CancellationToken cancellationToken = defa
         protected virtual IReadOnlyList<MigrationCommand> GetCreateTablesCommands(
             MigrationsSqlGenerationOptions options = MigrationsSqlGenerationOptions.Default)
             => Dependencies.MigrationsSqlGenerator.Generate(
-                Dependencies.ModelDiffer.GetDifferences(null, Dependencies.Model.GetRelationalModel()), Dependencies.Model, options);
+                Dependencies.ModelDiffer.GetDifferences(null, Dependencies.CurrentContext.Context.DesignTimeModel.GetRelationalModel()),
+                Dependencies.CurrentContext.Context.DesignTimeModel,
+                options);
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Storage/RelationalDatabaseCreator.cs:149]: @bricelam

File: src/EFCore/DbContext.cs [hunk 6]
```diff
@@ -138,11 +138,22 @@ public virtual ChangeTracker ChangeTracker
...
+        }
+
+        /// <summary>
+        ///     The metadata about the shape of entities, the relationships between them, and how they map to the database.
+        ///     Also includes all the information necessary to initialize the database.
+        /// </summary>
+        public virtual IModel DesignTimeModel
+        {
+            [DebuggerStepThrough]
+            get => ContextServices.DesignTimeModel;
```
Comment (AndriySvyryd) [inline src/EFCore/DbContext.cs:156]: @ajcvickers

File: src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                        {
+                            var slimParameter = Create(parameter, slimFunction);
+
+                            CreateAnnotations(parameter, slimParameter, static (convention, annotations, source, target, runtime) =>
+                                convention.ProcessFunctionParameterAnnotations(annotations, source, target, runtime));
+                        }
+
+                        CreateAnnotations(functionPair.Value, slimFunction, static (convention, annotations, source, target, runtime) =>
+                            convention.ProcessFunctionAnnotations(annotations, source, target, runtime));
+                    }
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs:76]: @smitpatel

File: src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+                    annotations[RelationalAnnotationNames.CheckConstraints] = slimCheckConstraints;
+                }
+
+                // These need to be set explicitly to prevent default values from being generated
+                annotations[RelationalAnnotationNames.TableName] = entityType.GetTableName();
+                annotations[RelationalAnnotationNames.Schema] = entityType.GetSchema();
+                annotations[RelationalAnnotationNames.ViewName] = entityType.GetViewName();
+                annotations[RelationalAnnotationNames.ViewSchema] = entityType.GetViewSchema();
+                annotations[RelationalAnnotationNames.SqlQuery] = entityType.GetSqlQuery();
+                annotations[RelationalAnnotationNames.FunctionName] = entityType.GetFunctionName();
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs:141]: @roji

File: src/EFCore/Infrastructure/ModelRuntimeInitializer.cs [hunk 9]
```diff
@@ -70,18 +73,35 @@ public ModelRuntimeInitializer(ModelRuntimeInitializerDependencies dependencies)
...
+                    static model =>
+                    {
+                        if (model is Model mutableModel)
+                        {
+                            model = mutableModel.OnModelFinalized();
+                        }
+
+                        return model!;
+                    },
+                    model);
```
Comment (AndriySvyryd) [inline src/EFCore/Infrastructure/ModelRuntimeInitializer.cs:104]: @roji

File: src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+        /// <param name="slimProperty"> The target property that will contain the annotations. </param>
+        /// <param name="runtime"> Indicates whether the given annotations are runtime annotations. </param>
+        protected override void ProcessPropertyAnnotations(
+            Dictionary<string, object?> annotations, IProperty property, SlimProperty slimProperty, bool runtime)
+        {
+            base.ProcessPropertyAnnotations(annotations, property, slimProperty, runtime);
+
+            if (runtime)
+            {
+                annotations.Remove(RelationalAnnotationNames.TableColumnMappings);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs:277]: Could we also remove column name/type ?

**Review by smitpatel — APPROVED** [review]
:shipit:

File: src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs (hunk 10 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs:277]: Unfortunately no, we need those to build the relational model and sometimes to get the type mapping

File: src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+            IDictionary<string, object?> annotations,
+            IKey key,
+            SlimKey slimKey,
+            bool runtime)
+        {
+            base.ProcessKeyAnnotations(annotations, key, slimKey, runtime);
+
+            if (runtime)
+            {
+                annotations.Remove(RelationalAnnotationNames.UniqueConstraintMappings);
```
[the remark this answers is not in the document]
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Conventions/RelationalSlimModelConvention.cs:342]: It's needed to find "shared" keys when building the relational model.

File: src/EFCore.Relational/Migrations/HistoryRepository.cs [hunk 12]
```diff
@@ -108,7 +108,7 @@ private IModel EnsureModel()
                         x.ToTable(TableName, TableSchema);
                     });
 
-                _model = Dependencies.ModelRuntimeInitializer.Initialize(modelBuilder.FinalizeModel(), validationLogger: null);
+                _model = Dependencies.ModelRuntimeInitializer.Initialize(modelBuilder.FinalizeModel(), designTime: true, validationLogger: null);
```
Comment (bricelam) [inline src/EFCore.Relational/Migrations/HistoryRepository.cs:111]: I'd love to rip all this code out... It was a mistake to try and mimic EF6 here.

File: test/EFCore.Relational.Specification.Tests/Migrations/MigrationsInfrastructureTestBase.cs [hunk 13]
```diff
@@ -307,12 +307,12 @@ public virtual void Can_get_active_provider()
         protected virtual void DiffSnapshot(ModelSnapshot snapshot, DbContext context)
         {
             var sourceModel = context.GetService<IModelRuntimeInitializer>().Initialize(
-                ((IMutableModel)snapshot.Model).FinalizeModel(), validationLogger: null);
+                ((IMutableModel)snapshot.Model).FinalizeModel(), designTime: false, validationLogger: null);
```
Comment (bricelam) [inline test/EFCore.Relational.Specification.Tests/Migrations/MigrationsInfrastructureTestBase.cs:310]: `true`?

**Review by bricelam — APPROVED** [review]

File: src/EFCore.InMemory/Storage/Internal/InMemoryStore.cs [hunk 14]
```diff
@@ -80,11 +81,13 @@ public class InMemoryStore : IInMemoryStore
 
                     var updateAdapter = updateAdapterFactory.CreateStandalone();
                     var entries = new List<IUpdateEntry>();
-                    foreach (var entityType in updateAdapter.Model.GetEntityTypes())
+                    foreach (var entityType in designModel.GetEntityTypes())
```
Comment (ajcvickers) [inline src/EFCore.InMemory/Storage/Internal/InMemoryStore.cs:84]: Why does this need to be the design model?

Comment (AndriySvyryd) [reply src/EFCore.InMemory/Storage/Internal/InMemoryStore.cs:84]: Semantically any code that creates the store needs to use the design model. In practice the slim model doesn't contain seed data.

File: src/EFCore/DbContext.cs (hunk 6 above)
Comment (ajcvickers) [reply src/EFCore/DbContext.cs:156]: We can argue about naming later...

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore/DbContext.cs (hunk 6 above)
Comment (smitpatel) [reply src/EFCore/DbContext.cs:156]: "argue" rather than "discuss" :trollface:

**Pull request merged**